tasks/ast_codegen/src/layout.rs

```diff
@@ -20,52 +24,63 @@ pub enum Layout {
+                HashMap::from([$((stringify!($typ), $layout_64)),*]);
+        }
+
+        lazy_static! {
+            static ref WELL_KNOWN_X32: HashMap<&'static str, Layout> =
+                HashMap::from([$((stringify!($typ), $layout_32)),*]);
+        }
     };
 }
 
-lazy_static! {
-    static ref WELL_KNOWN: HashMap<&'static str, Layout> = map! {
-        // builtins
-        bool => Layout::builtin::<bool>(),
-        u8 => Layout::builtin::<u8>(),
-        i8 => Layout::builtin::<i8>(),
-        u16 => Layout::builtin::<u16>(),
-        i16 => Layout::builtin::<i16>(),
-        u32 => Layout::builtin::<u32>(),
-        i32 => Layout::builtin::<i32>(),
-        u64 => Layout::builtin::<u64>(),
-        i64 => Layout::builtin::<i64>(),
-        u128 => Layout::builtin::<u128>(),
-        i128 => Layout::builtin::<i128>(),
-        f32 => Layout::builtin::<f32>(),
-        f64 => Layout::builtin::<f64>(),
-        char => Layout::builtin::<char>(),
-        &str => Layout::builtin::<&str>(),
-        // platform dependant
-        usize => Layout::Unknown,
-        isize => Layout::Unknown,
-        // well known types
-        // TODO: generate const assertions for these in the ast crate
-        Span => Layout::Layout { size: 8, align: 4 },
-        Atom => Layout::builtin::<&str>(),
-    };
+type Ptr32 = u32;
+
+well_known! {
+    // builtins
+    // types smaller or equal to 4 bytes are identical
+    bool => { 64: Layout::of::<bool>(), 32: Layout::of::<bool>()},
+    u8 => { 64: Layout::of::<u8>(), 32: Layout::of::<u8>() },
+    i8 => { 64: Layout::of::<i8>(), 32: Layout::of::<i8>() },
+    u16 => { 64: Layout::of::<u16>(), 32: Layout::of::<u16>() },
+    i16 => { 64: Layout::of::<i16>(), 32: Layout::of::<i16>() },
+    u32 => { 64: Layout::of::<u32>(), 32: Layout::of::<u32>() },
+    i32 => { 64: Layout::of::<i32>(), 32: Layout::of::<i32>() },
+    f32 => { 64: Layout::of::<f32>(), 32: Layout::of::<f32>() },
+    // alignment of these on 32bit machines are 4
+    u64 => { 64: Layout::of::<u64>(), 32: Layout::Layout { size: 8, align: 4 } },
+    i64 => { 64: Layout::of::<i64>(), 32: Layout::Layout { size: 8, align: 4 } },
+    f64 => { 64: Layout::of::<f64>(), 32: Layout::Layout { size: 8, align: 4 } },
+    u128 => { 64: Layout::of::<u128>(), 32: Layout::of::<u128>() },
+    i128 => { 64: Layout::of::<i128>(), 32: Layout::of::<i128>() },
+    char => { 64: Layout::of::<char>(), 32: Layout::of::<char>() },
+    &str => { 64: Layout::of::<&str>(), 32: Layout::of::<Ptr32>() },
+    // platform dependant
+    usize => { 64: Layout::of::<usize>(), 32: Layout::of::<u32>() },
+    isize => { 64: Layout::of::<isize>(), 32: Layout::of::<i32>() },
+    // well known types
+    // TODO: generate const assertions for these in the ast crate
+    Span => { 64: Layout::of::<oxc_span::Span>(), 32: Layout::of::<oxc_span::Span>() },
+    Atom => { 64: Layout::of::<oxc_span::Atom>(), 32: Layout::of::<oxc_span::Atom>() },
```

Comment:
I strongly believe we shouldn't extend the `ast` marker to types outside of the AST crate. Those one-off types can always be made `repr(C)` without all this infra. Span is already in the ideal order, And Vec type doesn't expose its fields so it can be reordered manually. Atom is a transparent type so no need for this stuff whatsoever.

But that's just my preference, I still haven't seen any real need to process those types.

Reply:
With that said, I have nothing against hard coding them as opposed to using a layout of method. I just did it for the sake of consistency if we regenerate on multiple architectures.

Reply:
For `Span` and `Atom` we can mark them `#[repr(C)]` and hard-code them if you prefer. But NB:

1. AST transfer needs to know the field order for `Span` (so we'd have to hard-code that too).
2. We will need to extend outside of `oxc_ast` crate, because there are various types in `oxc_syntax` crate which are part of the AST e.g. `AssignmentOperator`. These types need making `#[repr(C)]` and need explicit discriminants.

What's the reason for your opposition to codegen reaching into other crates?

Reply:
I've mulled over this, I guess you are right it would be easier if we do so.
I wanted to keep the code generation's input limited to AST types, Vec, Box, Span, and Atom obviously aren't AST types. But your argument for types defined in `oxc_syntax` is valid.

I'm not familiar with all AST-related stuff in the `oxc_syntax` crate,  But All I can remember is unit enums. So they should be easy to make `repr(C)`.

Initially, I was slightly biased against doing this but now I'm natural. It sure helps not to have to worry about those types. Even if they all are simple enums.
I'll do it in a follow-up PR.